src/main/java/net/snowflake/client/core/arrow/MapConverter.java

```diff
@@ -28,6 +31,30 @@ public Object toObject(int index) throws SFException {
 
   @Override
   public String toString(int index) throws SFException {
-    return vector.getObject(index).toString();
+    ArrowObjectStringRepresentationBuilder builder = new ArrowObjectStringRepresentationBuilder();
+
+    FieldVector vectorUnpacked = vector.getChildrenFromFields().get(0);
+
+    FieldVector keys = vectorUnpacked.getChildrenFromFields().get(0);
+    FieldVector values = vectorUnpacked.getChildrenFromFields().get(1);
+    final ArrowVectorConverter keyConverter;
+    final ArrowVectorConverter valueConverter;
+
+    SnowflakeType valueLogicalType =
+        ArrowVectorConverter.getSnowflakeTypeFromFieldMetadata(values.getField());
+
+    try {
+      keyConverter = ArrowVectorConverter.initConverter(keys, context, columnIndex);
+      valueConverter = ArrowVectorConverter.initConverter(values, context, columnIndex);
+    } catch (SnowflakeSQLException e) {
+      return vector.getObject(index).toString();
+    }
+
+    for (int i = vector.getElementStartIndex(index); i < vector.getElementEndIndex(index); i++) {
+      builder.appendKeyValue(
```

Comment:
What about keyLogicalType? I know that it could only String but it must be changed in future because database could return also Integers keys.

Reply:
It's used only to make the decision on whether the value should be quoted or not. The output string is JSON-like so the key is always quoted even for Integers